sdk/core/core-https/src/xhrHttpsClient.ts

```diff
@@ -59,31 +59,63 @@ export class XhrHttpsClient implements HttpsClient {
     for (const [name, value] of request.headers) {
       xhr.setRequestHeader(name, value);
     }
-    xhr.responseType = request.streamResponseBody ? "blob" : "text";
+    xhr.responseType = request.streamResponseStatusCodes?.size ?? 0 > 0 ? "blob" : "text";
 
     if (isReadableStream(request.body)) {
       throw new Error("Node streams are not supported in browser environment.");
     }
 
     xhr.send(request.body === undefined ? null : request.body);
 
-    if (request.streamResponseBody) {
+    if (request.streamResponseStatusCodes?.size ?? 0 > 0) {
       return new Promise((resolve, reject) => {
         xhr.addEventListener("readystatechange", () => {
           // Resolve as soon as headers are loaded
           if (xhr.readyState === XMLHttpRequest.HEADERS_RECEIVED) {
-            const blobBody = new Promise<Blob>((resolve, reject) => {
+            if (request.streamResponseStatusCodes?.has(xhr.status)) {
+              // eslint-disable-next-line @typescript-eslint/no-shadow
+              const blobBody = new Promise<Blob>((resolve, reject) => {
+                xhr.addEventListener("load", () => {
+                  resolve(xhr.response);
+                });
+                rejectOnTerminalEvent(request, xhr, reject);
+              });
+              resolve({
+                request,
+                status: xhr.status,
+                headers: parseHeaders(xhr),
+                blobBody
+              });
+            } else {
               xhr.addEventListener("load", () => {
-                resolve(xhr.response);
+                // response comes back in Blob when xhr.responseType === "blob"
+                // but the response body type is not expected to be stream based on response status code
+                // so converting from Blob to text
```

Comment:
I'm confused about this comment. How do we know the response is a blob and that we can't just read `xhr.responseText`? I don't see any actual detection logic, we're only asserting that this *isn't* a known streaming case.

Reply:
I assume that our operation spec lists all streaming operations thus we don't need any more detection logic.

- Before we send a request with xhr we set `xhr.responseType` to be either `blob` or `text` (default will be `text` if not specified). The response will come back in `response.blob` or `response.text` accordingly. https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/responseType
-  If the operation is streaming (based on our heuristic of any status code having a Stream body mapper) we use `blob` to send request.
  - The doc also mentions that "If the server returns data that is not compatible with the responseType that was set, the value of response will be null." but in my testing, it still worked when we send a request using `blob` and Storage service return the error response in `application/xml` in `response.blob`.
- So in the cases where we send request for streaming operations and response status code doesn't match any of the streaming status codes, the payload is stored in `response.blob` and needed to be converted into string.

Reply:
It turns out I was reading the code wrong. The if block was so big for if (streaming) that I didn't notice the else case (that was excluded from the diff view) where we use text in the non-blob case, so I thought we were treating **every** response as a blob.

Maybe as a refactoring we could move the blob case out into a helper function? That would also fix the nested-promise constructor naming collision issue.

Reply:
Good idea!  I extracted the "blob" case into a function.